From 1fc7e560a6689764ef8e0da1dfa5555bcc78cba6 Mon Sep 17 00:00:00 2001
From: Lennart Poettering <lennart@poettering.net>
Date: Mon, 14 Jan 2013 21:05:17 +0100
Subject: [PATCH 166/204] service: for Type=forking services, ignore exit
 status of main process depending on ExecStart's ignore setting

https://bugzilla.redhat.com/show_bug.cgi?id=860464
(cherry picked from commit fbeefb45ac1a257a0c5af975ad26d68ed6c39fda)
---
 src/core/service.c | 18 ++++++++++++++----
 1 file changed, 14 insertions(+), 4 deletions(-)

diff --git a/src/core/service.c b/src/core/service.c
index dbf3bea..6b453aa 100644
--- a/src/core/service.c
+++ b/src/core/service.c
@@ -2945,15 +2945,25 @@ static void service_sigchld_event(Unit *u, pid_t pid, int code, int status) {
                 s->main_pid = 0;
                 exec_status_exit(&s->main_exec_status, &s->exec_context, pid, code, status);
 
-                /* If this is not a forking service than the main
-                 * process got started and hence we copy the exit
-                 * status so that it is recorded both as main and as
-                 * control process exit status */
                 if (s->main_command) {
+                        /* If this is not a forking service than the
+                         * main process got started and hence we copy
+                         * the exit status so that it is recorded both
+                         * as main and as control process exit
+                         * status */
+
                         s->main_command->exec_status = s->main_exec_status;
 
                         if (s->main_command->ignore)
                                 f = SERVICE_SUCCESS;
+                } else if (s->exec_command[SERVICE_EXEC_START]) {
+
+                        /* If this is a forked process, then we should
+                         * ignore the return value if this was
+                         * configured for the starter process */
+
+                        if (s->exec_command[SERVICE_EXEC_START]->ignore)
+                                f = SERVICE_SUCCESS;
                 }
 
                 log_struct(f == SERVICE_SUCCESS ? LOG_DEBUG : LOG_NOTICE,
-- 
1.8.1.5

